x/farming/abci.go

```diff
@@ -3,21 +3,27 @@ package farming
 import (
 	"time"
 
-	abci "github.com/tendermint/tendermint/abci/types"
-
 	"github.com/cosmos/cosmos-sdk/telemetry"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 
 	"github.com/tendermint/farming/x/farming/keeper"
 	"github.com/tendermint/farming/x/farming/types"
 )
 
-// BeginBlocker distribute farming rewards for the previous block
-func BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock, k keeper.Keeper) {
-	defer telemetry.ModuleMeasureSince(types.ModuleName, time.Now(), telemetry.MetricKeyBeginBlocker)
-}
-
-func EndBlocker(ctx sdk.Context, k keeper.Keeper) []abci.ValidatorUpdate {
+func EndBlocker(ctx sdk.Context, k keeper.Keeper) {
 	defer telemetry.ModuleMeasureSince(types.ModuleName, time.Now(), telemetry.MetricKeyEndBlocker)
-	return []abci.ValidatorUpdate{}
+
+	lastEpochTime, found := k.GetLastEpochTime(ctx)
+	if !found {
+		k.SetLastEpochTime(ctx, ctx.BlockTime())
```

Comment:
some cases in sdk using `ctx.BlockHeader().Time` rather than `ctx.BlockTime()`, Can you check if there is any difference between these two methods?

Reply:
As you can find in: https://github.com/cosmos/cosmos-sdk/blob/8f7be3e1554b831501ec40d5aca0a3ae62468473/types/context.go#L50-L66

```go
func (c Context) BlockTime() time.Time        { return c.header.Time }
...
// clone the header before returning
func (c Context) BlockHeader() tmproto.Header {
	var msg = proto.Clone(&c.header).(*tmproto.Header)
	return *msg
}
```

`BlockHeader()` returns a clone of the header. Accessing the `Header.Time` field will return same timestamp, but `BlockTime()` seems much more efficient way for doing that.